### PR TITLE
AM-1227: manage mfa

### PR DESCRIPTION
**Pull request by mstasgravitee**
## :id: Reference related issue. 
https://gravitee.atlassian.net/browse/AM-1227

### CODE REVIEW

File: gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchMFASettings.java
Comment (leleueri) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchMFASettings.java:1]: **issue**: the stepUpAuthentication attribute added in the MFASettings is not present in the Patch bean (in previous PR). Is it normal ?

File: gravitee-am-ui/src/app/domain/applications/application/advanced/factors/factors.component.ts [hunk 1]
```diff
@@ -79,40 +90,43 @@ export class ApplicationFactorsComponent implements OnInit {
...
     const data = {
       factors: this.application.factors,
       settings: {
         riskAssessment: this.riskAssessment,
         mfa: {
           stepUpAuthenticationRule: this.mfaStepUpRule,
+          // stepUpAuthentication: {
+          //   active: true,
+          //   stepUpAuthenticationRule: this.mfaStepUpRule,
+          // },
```
Comment (leleueri) [inline gravitee-am-ui/src/app/domain/applications/application/advanced/factors/factors.component.ts]: **question**: I am confused here, which stepUp attribute is the one to use starting from 4.3. stepUpAuthenticationRule or the stepUpAuthentication?

File: gravitee-am-ui/src/app/domain/applications/application/advanced/factors/factors.component.ts [hunk 2]
```diff
@@ -28,24 +29,32 @@ import { AuthService } from '../../../../../services/auth.service';
...
   rememberDevice: any = {};
-
   adaptiveMfaRule = '';
+  skipAdaptiveMfaRule = '';
   enrollment: any = {};
   selectedMFAOption = '';
-  private riskAssessment: any = {};
+  selectedChallengeMFAOption = '';
+  challengeActive: boolean;
+  skipChallengeRule: string;
```
Comment (leleueri) [inline gravitee-am-ui/src/app/domain/applications/application/advanced/factors/factors.component.ts]: **thought**: Why did you add dedicated attribute here (ex: challengeActive & skipChallengeRule) instead of adding a Challenge structure as we have in the Java part ? It maybe more simple to follow if the FrontEnd and the Backend settings structures are the same IMO. wdyt ?

File: gravitee-am-model/src/main/java/io/gravitee/am/model/MFASettings.java [hunk 3]
```diff
@@ -18,30 +18,35 @@
...
+@Setter
+@NoArgsConstructor
 public class MFASettings {
 
     private String loginRule;
     //deprecated since 4.3. use stepUpAuthentication instead
     @Deprecated
     private String stepUpAuthenticationRule;
     private StepUpAuthenticationSettings stepUpAuthentication;
+    private String skipStepUpAuthentication;
```
[the remark this answers is not in the document]
Comment (mstasgravitee) [reply gravitee-am-model/src/main/java/io/gravitee/am/model/MFASettings.java]: it was removed and added by mistake

File: gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchMFASettings.java
Comment (mstasgravitee) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchMFASettings.java:1]: it was removed and added by mistake

File: gravitee-am-model/src/main/java/io/gravitee/am/model/MFASettings.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mstasgravitee) [reply gravitee-am-model/src/main/java/io/gravitee/am/model/MFASettings.java]: it was removed and added by mistake.

File: gravitee-am-ui/src/app/domain/applications/application/advanced/factors/factors.component.ts (hunk 1 above)
Comment (mstasgravitee) [reply gravitee-am-ui/src/app/domain/applications/application/advanced/factors/factors.component.ts]: Outdated

File: gravitee-am-ui/src/app/domain/applications/application/advanced/factors/factors.component.ts (hunk 2 above)
Comment (mstasgravitee) [reply gravitee-am-ui/src/app/domain/applications/application/advanced/factors/factors.component.ts]: Outdated. This is an output from the challenge component and the challenge structure is different than in the backend.

File: gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchMFASettings.java
Comment (mstasgravitee) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/model/PatchMFASettings.java:1]: yes, it is missing. Added.

File: gravitee-am-ui/src/app/domain/applications/application/advanced/factors/mfa-challenge/mfa-challenge.component.ts [hunk 4]
```diff
@@ -1,15 +1,183 @@
...
+    if (isRiskBased) {
+      this.adaptiveMfaRule = this.computeRiskAssessmentRule(this.riskAssessment);
+    } else if (this.currentMode === MfaChallengeComponent.modeOptions.CONDITIONAL) {
+      this.adaptiveMfaRule = this.challenge.challengeRule;
+    } else {
+      this.adaptiveMfaRule = '';
+      this.challenge.challengeRule = '';
+    }
+    this.settingsChange.emit({
+      active: true,
```
Comment (leleueri) [inline gravitee-am-ui/src/app/domain/applications/application/advanced/factors/mfa-challenge/mfa-challenge.component.ts:176]: **issue**: this boolean shouldn't be force to true but should reflect the user decision when he enabled or disabled the toggle.

File: gravitee-am-ui/src/app/domain/applications/application/advanced/factors/mfa-challenge/risk-based/mfa-risk-based.component.html [hunk 5]
```diff
@@ -15,26 +15,30 @@
     limitations under the License.
 
 -->
-
+<info-banner
```
Comment (leleueri) [inline gravitee-am-ui/src/app/domain/applications/application/advanced/factors/mfa-challenge/risk-based/mfa-risk-based.component.html:18]: **question**: When I click on the "More Info" button a new tab is open to see the details but the application settings are saved when we click on this button. Is it the right behaviour ? I have some concerns about this behaviour as if the application settings are not defined but not finalized, the save action will redeploy the application on the gateway with unexpected settings... this is an issue IMO.

**Review by leleueri — CHANGES_REQUESTED** [review]

File: gravitee-am-model/src/main/java/io/gravitee/am/model/MfaType.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.am.model;
+
+/**
+ * @author Ashraful HASAN (ashraful.hasan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public enum MfaType {
```
[the remark this answers is not in the document]
Comment (mstasgravitee) [reply gravitee-am-model/src/main/java/io/gravitee/am/model/MfaType.java:22]: yes we can create 2 types

File: gravitee-am-ui/src/app/domain/applications/application/advanced/factors/factors.component.ts [hunk 7]
```diff
@@ -60,59 +67,72 @@ export class ApplicationFactorsComponent implements OnInit {
...
       settings: {
-        riskAssessment: this.riskAssessment,
+        riskAssessment: this.challenge.riskAssessment,
         mfa: {
           stepUpAuthenticationRule: this.mfaStepUpRule,
-          adaptiveAuthenticationRule: this.adaptiveMfaRule,
+          stepUpAuthentication: {
+            active: this.mfaStepUpRule !== '',
+            stepUpAuthenticationRule: this.mfaStepUpRule,
+          },
```
[the remark this answers is not in the document]
Comment (mstasgravitee) [reply gravitee-am-ui/src/app/domain/applications/application/advanced/factors/factors.component.ts:110]: yes, it is not a final solution. The frontend component is not changed for that part yet. This will be changed in the next tasks.

File: gravitee-am-model/src/main/java/io/gravitee/am/model/StepUpAuthenticationSettings.java [hunk 8]
```diff
@@ -15,35 +15,23 @@
...
 /**
  * @author Ashraful HASAN (ashraful.hasan at graviteesource.com)
  * @author GraviteeSource Team
  */
+@Getter
+@Setter
+@NoArgsConstructor
 public class StepUpAuthenticationSettings {
     private Boolean active;
     private String stepUpAuthenticationRule;
```
[the remark this answers is not in the document]
Comment (mstasgravitee) [reply gravitee-am-model/src/main/java/io/gravitee/am/model/StepUpAuthenticationSettings.java:31]: We can rename it later after we merge the @ashraf706's branch to mitigate conflicts.

File: gravitee-am-ui/src/app/domain/applications/application/advanced/factors/remember-device/mfa-remember-device.component.html [hunk 9]
```diff
@@ -25,71 +25,79 @@
...
-      [disabled]="!isActive()"
-      (change)="onSkip($event)"
-      [checked]="skipRememberDevice"
+  <div class="config-section">
+    <info-banner
+      *ngIf="!hasDeviceIdentifierPlugins()"
+      [link]="getDeviceIdentifierLink()"
+      text="To use the Remember Device feature, you need to first create a Device Identifier Plugin."
+      type="warning"
+      buttonName="Add device identifier"
```
[the remark this answers is not in the document]
Comment (mstasgravitee) [reply gravitee-am-ui/src/app/domain/applications/application/advanced/factors/remember-device/mfa-remember-device.component.html:45]: fixed

File: gravitee-am-ui/src/app/domain/applications/application/advanced/factors/mfa-challenge/risk-based/mfa-risk-based.component.html (hunk 5 above)
Comment (mstasgravitee) [reply gravitee-am-ui/src/app/domain/applications/application/advanced/factors/mfa-challenge/risk-based/mfa-risk-based.component.html:18]: fixed

File: gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/internal/model/ChallengeSettingsMongo.java
[the remark this answers is not in the document]
Comment (mstasgravitee) [reply gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/internal/model/ChallengeSettingsMongo.java:1]: added

File: gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/internal/model/EnrollSettingsMongo.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    private Long skipTimeSeconds;
+    private String enrollmentRule;
+    private String type;
+
+    public static EnrollSettingsMongo convert(EnrollSettings enrollment) {
+        return Optional.of(enrollment).filter(Objects::nonNull).map(settings -> {
+            var enrollmentMongo = new EnrollSettingsMongo();
+            enrollmentMongo.setActive(TRUE.equals(settings.isActive()));
+            enrollmentMongo.setForceEnrollment(settings.getForceEnrollment());
+            enrollmentMongo.setSkipTimeSeconds(settings.getSkipTimeSeconds());
```
Comment (leleueri) [inline gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/internal/model/EnrollSettingsMongo.java:42]: **question**: could it be the same issue as the Active boolean ?

File: gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/internal/model/EnrollSettingsMongo.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+            enrollmentMongo.setType(settings.getType() != null ? settings.getType().name() : null);
+            return enrollmentMongo;
+        }).orElse(new EnrollSettingsMongo());
+    }
+
+    public EnrollSettings convert() {
+        var enrollmentSettings = new EnrollSettings();
+        enrollmentSettings.setActive(TRUE.equals(getActive()));
+        enrollmentSettings.setForceEnrollment(getForceEnrollment());
+        enrollmentSettings.setSkipTimeSeconds(getSkipTimeSeconds());
```
Comment (leleueri) [inline gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/internal/model/EnrollSettingsMongo.java:53]: **question**: could it be the same issue as the Active boolean ?

File: gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/internal/model/EnrollSettingsMongo.java (hunk 10 above)
Comment (mstasgravitee) [reply gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/internal/model/EnrollSettingsMongo.java:42]: No, it is not. Long can be null. It is copied from Deprecated EnrollmentSettings.

File: gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/internal/model/EnrollSettingsMongo.java (hunk 11 above)
Comment (mstasgravitee) [reply gravitee-am-repository/gravitee-am-repository-mongodb/src/main/java/io/gravitee/am/repository/mongodb/management/internal/model/EnrollSettingsMongo.java:53]: No, it is not. Long can be null. It is copied from Deprecated EnrollmentSettings.

**Review by leleueri — APPROVED** [review]
